api/profile.go

```diff
@@ -54,123 +49,60 @@ func (h *ProfileHandlers) getProfileHandler(w http.ResponseWriter, r *http.Reque
 	util.WriteResponse(w, res)
 }
 
-func (h *ProfileHandlers) saveProfileHandler(w http.ResponseWriter, r *http.Request) {
-	p := &config.ProfilePod{}
-	if err := json.NewDecoder(r.Body).Decode(p); err != nil {
-		util.WriteErrResponse(w, http.StatusBadRequest, fmt.Errorf("error decoding request body: %s", err.Error()))
-		return
-	}
-	res := &config.ProfilePod{}
-	if err := h.SaveProfile(p, res); err != nil {
-		util.WriteErrResponse(w, http.StatusInternalServerError, fmt.Errorf("error saving profile: %s", err.Error()))
-		return
-	}
-	util.WriteResponse(w, res)
-}
-
 // ProfilePhotoHandler is the endpoint for uploading this peer's profile photo
 func (h *ProfileHandlers) ProfilePhotoHandler(w http.ResponseWriter, r *http.Request) {
 	switch r.Method {
-	case http.MethodGet:
+	case http.MethodGet, http.MethodPost:
 		h.getProfilePhotoHandler(w, r)
-	case http.MethodPut, http.MethodPost:
-		h.setProfilePhotoHandler(w, r)
 	default:
 		util.NotFoundHandler(w, r)
 	}
 }
 
 func (h *ProfileHandlers) getProfilePhotoHandler(w http.ResponseWriter, r *http.Request) {
-	data := []byte{}
-	req := &config.ProfilePod{}
-	req.Peername = r.FormValue("peername")
-	req.ID = r.FormValue("id")
-
-	if err := h.ProfilePhoto(req, &data); err != nil {
-		util.WriteErrResponse(w, http.StatusInternalServerError, err)
+	params := &config.ProfilePod{}
+	err := lib.UnmarshalParams(r, params)
+	if err != nil {
+		util.WriteErrResponse(w, http.StatusBadRequest, err)
 		return
 	}
 
-	w.Header().Set("Content-Type", "image/jpeg")
-	w.Write(data)
-}
-
-func (h *ProfileHandlers) setProfilePhotoHandler(w http.ResponseWriter, r *http.Request) {
-	p := &lib.FileParams{}
-	if r.Header.Get("Content-Type") == "application/json" {
-		json.NewDecoder(r.Body).Decode(p)
-	} else {
-		infile, header, err := r.FormFile("file")
-		if err != nil && err != http.ErrMissingFile {
-			util.WriteErrResponse(w, http.StatusBadRequest, err)
-			return
-		}
-
-		p = &lib.FileParams{
-			// Url:          r.FormValue("url"),
-			Filename: header.Filename,
-			Data:     infile,
-		}
-	}
-
-	res := &config.ProfilePod{}
-	if err := h.SetProfilePhoto(p, res); err != nil {
-		log.Infof("error initializing dataset: %s", err.Error())
-		util.WriteErrResponse(w, http.StatusInternalServerError, err)
+	res, err := h.ProfilePhoto(r.Context(), params)
+	if err != nil {
+		log.Infof("error getting profile photo: %s", err.Error())
+		util.WriteErrResponse(w, http.StatusBadRequest, err)
 		return
 	}
-	util.WriteResponse(w, res)
+
+	w.Header().Set("Content-Type", "image/jpeg")
```

Comment:
I wonder if we're going to soon want some like UnmarshalParams but for outgoing results. Some kind of hook so that methods can add additional outgoing information.

Reply:
Yeah, had similar thought, but think we're still need to refactor on step deeper as several methods conflate different types of output and that feels like we would just move bloat around. If we were to go that route, I'd first like to separate purely JSON APIs from raw data endpoints and work from there.